### ISSUE TITLE
Add __get__ support for descriptors

### ISSUE THREAD

**Issue by mcstarioni** [theY4Kman/pycharm-typing-imp#1]
Thanks for you civic activity and contribution to what should be done by JetBrains, but could you also add support for `__get__` method typing annotations? [Here ](https://youtrack.jetbrains.com/issue/PY-55531/Pycharm-cant-handle-typing-of-get-method-of-descriptors) is the link to the issue in YouTrack. Knowing how fast JetBrains deals with this type of issues, your plugin is the only hope.

**Comment by theY4Kman** [comment]
Hmm, that doesn't seem too difficult, at least in simplest case, where `__get__` is called with an instance. Determining different return types based on whether instance is `None` or not seems a bit more involved.

**Comment by theY4Kman** [comment]
I replied in that issue with a clarifying question. I had assumed _all_ descriptor `__get__()` accesses were being typed incorrectly, but it seems the typing issue is limited to `self` TypeVar substitutions. I'll see what I can do about that if so.

<details>
<summary>Issue reply reproduced here, for posterity</summary>

---

I'm trying to slim this down to a test case/minimally-reproducible example, because I'm not quite sure what one _should_ expect the type of `Container().b.var` to be. I'm also not sure `__get__` , in general, is faulty, because the following two cases pass in testing:

```py
class Attr:
    def __get__(self, instance, owner) -> float: ...

class Foo:
    my_attr = Attr()

foo = Foo()
foo.my_attr  # type is "float"
```

```py
class Attr:
    def __get__(self, instance, owner) -> float: ...

class AttrChild(Attr):
    def __get__(self, instance, owner) -> int: ...

class Foo:
    my_attr = AttrChild()

foo = Foo()
foo.my_attr  # type is "int"
```

Is the issue here is that the TypeVar on `self` isn't specializing to the descriptor class? Because I _am_ observing that the following two cases aren't substituting the descriptor class in the TypeVar:
```py
from typing import TypeVar

AttrT = TypeVar('AttrT', bound='Attr')

class Attr:
    def __get__(self: AttrT, instance, owner) -> AttrT: ...

class Foo:
    my_attr = Attr()

foo = Foo()
foo.my_attr  # expected "Attr", actual "AttrT"
```

```py
from typing import TypeVar

AttrT = TypeVar('AttrT', bound='Attr')

class Attr:
    def __get__(self: AttrT, instance, owner) -> AttrT: ...

class AttrChild(Attr):
    pass

class Foo:
    my_attr = AttrChild()

foo = Foo()
foo.my_attr  # expected "AttrChild", actual "AttrT"
```

For control purposes, I did test whether a regular method using a `TypeVar` on `self` gets substituted correctly, and found the behaviour correct:
```py
from typing import TypeVar

T = TypeVar('T')

class Foo:
    def my_fun(self: T) -> T: ...

foo = Foo()
rval = foo.my_fun()
rval  # type is "Foo"
```

So, is it more accurate to say `self` TypeVars aren't appropriately substituted on descriptors?

</details>

**Comment by theY4Kman** [comment]
Ayyyyyyy, I was able to find a reasonable solution, and I've published version 0.2.0 to the JB Marketplace; it'll be available once (and if?) they approve the update.

If you're impatient, the new version can be manually installed using the .zip from [the GH release](https://github.com/theY4Kman/pycharm-typing-imp/releases/tag/0.2.0).

Lemme know if it does or doesn't solve your issue! I'm gonna close this issue for now; if it doesn't work for you, I'll reopen.

Thanks for reporting! Keep 'em coming ;)
